### PR TITLE
Split up monolithic SConstruct into many small parts.

### PR DESCRIPTION
**Pull request by rec**
* New Module system.
  * Many modules to deal with common libraries.
  * Simple "hello, world" for Beast.

### CODE REVIEW

File: src/beast/python/beast/build/module/Boost.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import os
+from beast.build import Function, Module
+
+def boost(variant, link_libraries):
+    link_libraries = [i if i.startswith('boost_') else 'boost_' + i
+                      for i in link_libraries]
+    try:
+        root = variant.state.environ['BOOST_ROOT']
+    except KeyError:
+        raise KeyError('The environment variable BOOST_ROOT must be set')
```
Comment (seelabs) [inline src/beast/python/beast/build/module/Boost.py]: The old version did not raise an error when `BOOST_ROOT` was not present. This is causing problems with travis. We should also update our build docs if we keep this behavior.

File: bin/python/ripple/build/Config.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# This file is part of Beast: http://github.com/vinniefalco/Beast
+#
+# This file copyright (c) 2015, Tom Ritchford <tom@swirly.com>
+# under the Boost software license http://www.boost.org/LICENSE_1_0.txt
```
Comment (seelabs) [inline bin/python/ripple/build/Config.py]: The copyright should probably be `ripple` instead of `Tom Ritchford`

**Comment by seelabs** [comment]
A no-op build went from ~1 sec with the old sconstruct to ~4 secs with the new. We should look at putting some of those optimizations back. (I don't know what specifically yet).

File: SConstruct [hunk 3]
```diff
@@ -60,14 +64,10 @@ The following environment variables modify the build environment:
     OPENSSL_ROOT
       Path to the openssl directory.
 
-The following extra options may be used:
-    --ninja         Generate a `build.ninja` build file for the specified target
-                    (see: https://martine.github.io/ninja/). Only gcc and clang targets
-                     are supported.
+"""
```
Comment (seelabs) [inline SConstruct]: The `--fast` option needs to be documented in this file.

File: src/beast/python/beast/build/TagSet.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            tags = ', '.join(tags)
+            raise ValueError('Unknown tag%s: %s.' % (plural, tags))
+        return itertools.product(*product)
+
+    def __repr__(self):
+        return 'TagGroups(%s)' % repr(self.groups)
+
+
+class Targets(object):
+    """Map a command line to targets and TagGroupss.
```
Comment (seelabs) [inline src/beast/python/beast/build/TagSet.py]: Typo `TagGroupss`

File: src/beast/python/beast/build/Module.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+# This file is part of Beast: http://github.com/vinniefalco/Beast
+#
+# This file copyright (c) 2015, Tom Ritchford <tom@swirly.com>
+# under the Boost software license http://www.boost.org/LICENSE_1_0.txt
+
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals)
+
+import glob, os
+from functools import wraps
```
Comment (seelabs) [inline src/beast/python/beast/build/Module.py]: Is `wraps` ever used?

File: bin/python/ripple/build/Config.py (hunk 2 above)
Comment (vinniefalco) [inline bin/python/ripple/build/Config.py]: Also the file isn't part of beast its part of rippled...

File: bin/python/ripple/build/Unity.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# This file is part of Beast: http://github.com/vinniefalco/Beast
```
Comment (vinniefalco) [inline bin/python/ripple/build/Unity.py]: Not part of beast

File: src/beast/python/beast/build/module/Boost.py (hunk 1 above)
Comment (rec) [inline src/beast/python/beast/build/module/Boost.py]: Thanks, good catch!

File: bin/python/ripple/build/Config.py (hunk 2 above)
Comment (rec) [inline bin/python/ripple/build/Config.py]: OOPS!  :-)

File: src/beast/python/beast/build/TagSet.py (hunk 4 above)
Comment (rec) [inline src/beast/python/beast/build/TagSet.py]: My precioussss!

File: src/beast/python/beast/build/Module.py (hunk 5 above)
Comment (rec) [inline src/beast/python/beast/build/Module.py]: Hah, not any more!

Gone.

Comment (rec) [inline src/beast/python/beast/build/Module.py]: And neither were `glob`s nor `os`.

File: SConstruct (hunk 3 above)
Comment (rec) [inline SConstruct]: We're going to be having at least two SConstructs now - should this really go in both SConstructs?

Honestly, I'd like to get rid of all these comments from here and put them into an actual document.  (Also, I'd like to consider bringing back ninja mode, sorry I dropped that when I was struggling...)

File: bin/python/ripple/build/module/Default.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+Prepend = Env.Prepend
+
+
+MODULE = Module(
+    files=compose(
+        for_tags(
+            'debug',
+            Append(CPPDEFINES=['DEBUG', '_DEBUG']),
+        ),
+
```
Comment (seelabs) [inline bin/python/ripple/build/module/Default.py]: @rec: I really like the interface you have here. Very nice. However, I don't like the complexity needed to support this interface. I can't do better, but I'm also concerned about anyone on the team other than you supporting this python code. Looking at the implementation, there are lots of functions passed around. It makes the code hard to follow for someone (like me) who isn't used to the style. For example:

1) Look at what `for_tags` does. This is defined in `Function.py`

2) That calls: `_tagger` with the items and a matcher.

3) `_tagger` returns a function that iterates through all the callable objects that pass the matcher with the non-callable objects (I did a double-take on `(functions.append if callable(i) else tags.add)(i)`)

4) `Append` is `Env.Append`. Env is defined in `Function.py` as `EnvClass`. `EnvClass` doesn't have an attribute `Append`, but it does define a `__getattr__` function that return a function that returns a function (two levels) that calls a function on the build env.

5) `compose` is defined in `Function.py` and returns a function that calls all of the arguments originally passed to `compose`.

6) The `Module` `__init__` function looks at the `files` parameter. If it's callable, it saves it, if not there are other special cases.

I'm also having some problems navigating all the python files that support our build system.

I don't know how actionable that feedback is, but I did want to share the concern.

Comment (vinniefalco) [inline bin/python/ripple/build/module/Default.py]: What does "for_tags" mean, grammatically? Is that like a for_loop? What's a "for"? Whats "tags"?

File: src/beast/python/beast/build/BuildState.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            self.env.Alias(key, value)
+
+        module.run('after', self)
+
+
+DEFAULTS = {
+    'build_dir': 'build',
+
+    'environment': {
+        'toolpath': [os.path.join('src', 'beast', 'python', 'site_tools')],
```
Comment (seelabs) [inline src/beast/python/beast/build/BuildState.py]: Setting the path like this assumes that the SConstruct file lives in the root rippled directory. Of course, that is true for ripple, but it makes it difficult to re-use these beast python files for projects that do not live in the root. One solution would be to assume the `site_tools` directory lives in a path relative to this file (using `os.path.abspath(__file__)` to get that path to this file).

I ran across this while trying to build a simple "hello world" project to test some of this machinery. There may be other places where paths are set assuming a certain location for SConstruct, I'd like to fix thoses as well.

Comment (seelabs) [inline src/beast/python/beast/build/BuildState.py]: This does what I had in mind: https://github.com/seelabs/rippled/blob/53aaba32c4b32e10f545126ce25d073ecfffdc33/src/beast/python/beast/build/BuildState.py#L61-71

File: SConstruct (hunk 3 above)
Comment (seelabs) [inline SConstruct]: It isn't clear how to do common tasks with this SConstruct - for example, that most of the things a user wants to edit live in `bin/python/ripple/build/module/__init__.py`. I went through all the commits made to SConstruct since 2014 and made a list of common tasks. We should look at that list and make sure there's a doc explaining how to do the common tasks (unless it's obvious how to do them). Doc:  https://ripplelabs.atlassian.net/wiki/display/~determan/Typical+SConstruct+Tasks

File: src/beast/python/beast/build/BuildState.py (hunk 8 above)
Comment (rec) [inline src/beast/python/beast/build/BuildState.py]: Relative paths!  Good idea, I'll recast it all in that fashion.

File: bin/python/ripple/build/module/Default.py (hunk 7 above)
Comment (rec) [inline bin/python/ripple/build/module/Default.py]: I'm _very_ open to another name - I've gone through various incarnations of it over development!

A tag is a property of a build like `gcc`, `debug`, `unity` or `osx`.

The full sentence would be:  "For builds with tags [list of tags], do [this thing]" - `for_tags`.

Earlier incarnations were `Tags`, `with_tags`, `do_tags`


Tags are documented [here](https://github.com/rec/rippled/blob/25928e124567d9c01eeccee4ac7e5cd137ec0878/src/beast/python/beast/build/TagSet.py#L12) though I could probably revisit that.

Comment (rec) [inline bin/python/ripple/build/module/Default.py]: *Complexity*

Now, these are excellent points, each of which is an additional complexity
- and as you know I consider complexity a huge threat. (I can also fix some
of these issues...)

That said, there is a great deal of complexity in what we are requiring of
our build system, all of which has to go somewhere.


*Four phase build.*

Let's start with the "four phase build"
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/Module.py#L14-L26>
-
before, files, target, after.

We have main targets like rippled, and within them build variants like
gcc.debug and this structure is an existing precondition that can't be
changed.

This requires us to have two different build states:  we have one BuildState
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/BuildState.py>,
but multiple BuildVariants
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/BuildVariant.py>
.

Given the constraints, there isn't another way to do it.


*Functional programming*

Now, why are we returning functions everywhere?  Why not just... do things?

Two reasons:

The first is that Scons tends to work that way - you prepare a description
of what's going to happen, and then Scons does it.

The second is that I wanted Modules, as much as possible, to be expressible
as static data - something that can be constructed at program loading time
and never changed - because this makes reuse and reference so much easier.
But at the time the Python program loads, we simply don't have any build
information ready!

(Having separate Modules we can reuse was one of the stated goals of the
project...)



*Hard to navigate directories.*

As for the complexity of having code split over two separate directories
without a common root - tell me about it.  :-(  This is an unavoidable
consequence of having the segmentation between "beast" and "ripple".

I tried to make the hierarchies exactly the same:


    bin/python/ripple
    ├── build
    │   ├── Config.py
    │   ├── module
    │   │   ├── Default.py
    │   │   └── __init__.py
    │   └── Unity.py

    src/beast/python/beast
    ├── build
    │   ├── BuildState.py
    │   ├── BuildVariant.py
    │   ├── CcWarning.py
    │   ├── CheckForFastBuild.py
    │   ├── Config.py
    │   ├── Function.py
    │   ├── Git.py
    │   ├── module
    │   │   ├── Beast.py
    │   │   ├── Boost.py
    │   │   ├── CountTests.py
    │   │   ├── GitTag.py
    │   │   ├── Jemalloc.py
    │   │   ├── MSVC.py
    │   │   ├── OpenSSL.py
    │   │   ├── PhonyTargets.py
    │   │   ├── Proto.py
    │   │   └── Vcxproj.py
    │   ├── Module.py
    │   ├── TagSet.py
    │   ├── Toolchain.py
    │   └── Util.py
    ├── Color.py
    └── System.py


(This graph omits test code, and empty __init__.py files.)



*Maintenance*

Big issue for me!

What is likely to happen in future maintenance?   Here's my guess, by
decreasing frequency:

   1. We're going to add a lot of Modules.
   2. We'll add new binary targets: e.g. tests.
   3. We'll add new options like --help (or replacing --ninja).
   4. We'll occasionally have to add a new "build phase", or fix details in
   an existing build phase.
   5. We'll make existential overhauls to the whole "tags and variants"
   idea.


So a top priority was for me *making Modules easy to write and maintain.*

And - I think - they are.

It's really easy to create a module that just has source code
<https://github.com/rec/rippled/blob/scons/bin/python/ripple/build/module/__init__.py#L33-L38>
.

There's a special case
<https://github.com/rec/rippled/blob/scons/bin/python/ripple/build/module/__init__.py#L89-L106>
for unity builds if the unity file has the same name as the directory - but
we can also handle complex cases like the "legacy build"
<https://github.com/rec/rippled/blob/scons/bin/python/ripple/build/module/__init__.py#L43-L64>
.


Now, some of our modules have fairly complex logic.  For example, for
boost, there's a list of hard-compiled libraries, and we don't require
BOOST_ROOT to be set, but if it is, we need to explicitly load from there,
and to prefer the static libraries if possible - here
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/module/Boost.py>
.

Comment (rec) [inline bin/python/ripple/build/module/Default.py]: Second part.

Regarding Function and Env specifically... these do use "advanced" code to
implement, but I don't think we'll *ever* need to change it.

Client code that uses it is short
<https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/module/Beast.py#L15-L26>
and, I hope, understandable.

It is very rare that I actually import names unqualified - but at the end
of this I had it down to a tiny number of symbols that aren't likely to
grow much.


I'm actually going to go and cut these down even a little right now.  :-)

Comment (rec) [inline bin/python/ripple/build/module/Default.py]: Regarding relative imports:

http://stackoverflow.com/questions/14132789/python-relative-imports-for-the-billionth-time

I'm still digesting this...

Comment (vinniefalco) [inline bin/python/ripple/build/module/Default.py]: @rec okay, `for_tags` makes more sense now

**Comment by rec** [comment]
All right!

I made some more changes to clean things up though I still haven't fixed the problems I listed before:

In particular, a new [Build.py](https://github.com/rec/rippled/blob/scons/src/beast/python/beast/build/Build.py) has _all_ the names that a client uses from the build system.

**Comment by rec** [comment]
I just pushed a lot more changes.

Now it works perfectly again on OS/X.  I simplified a lot of names and some of the logic.

I still don't know if it works on Windows, anyone want to test for me?

**Comment by vinniefalco** [comment]
Does not work on Windows:
- The vcxproj is invalid
- `scons vcxproj` malfunctions

**Comment by rec** [comment]
Can I get more information than "malfunctions"?

**Comment by rec** [comment]
(I can fix the vcxproj if it doesn't work, but what goes wrong with the main build?)

**Comment by rec** [comment]
`scons vcxproj` works on Linux, and does not result in a delta...

**Comment by rec** [comment]
Fixing travis for this branch...

**Comment by vinniefalco** [comment]
@rec The vcxproj at the tip of the branch fails to load in msvc, and issuing `scons vcxproj` returns immediately with an empty vcxproj.

**Comment by rec** [comment]
OK, I can repro that.  I'll see what I can do!

File: src/beast/python/beast/build/State.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            self.env.Alias(key, value)
+
+        module.after(self)
+
+
+DEFAULTS = {
+    'build_dir': 'build',
+
+    'environment': {
+        'toolpath': [os.path.join('src', 'beast', 'python', 'site_tools')],
```
Comment (seelabs) [inline src/beast/python/beast/build/State.py]: @rec: I need two changes to build my "hello world" project:

1) Change this to a relative path. What I had in mind:  https://github.com/seelabs/rippled/blob/53aaba32c4b32e10f545126ce25d073ecfffdc33/src/beast/python/beast/build/BuildState.py#L61-71 

2) Resolve the warning about "Two different environments were specified for target main.o"

(apologies if this is already on your to-do list. I left a comment on an outdated diff and I wanted to make sure this didn't get lost).

**Comment by rec** [comment]
Update:  I got past two problems involving the vcxproj but now I have an issue which I just discovered was because I have a list of files, some of which are strings and some, SCons.Node.FS.File objects, and the VSProject code doesn't like that.  (Clearly the rest of Scons doesn't mind at all...)

I'm not stuck and we're getting further into it, just wanted to report.


It took me a while to realize overall how the vcxproj process works - in the special case of the target vcxproj, it adds targets as if for the msvc targets, and _as a side effect_ gathers some values for msvc in a "global variable".  The issue is that I had equated _adding_ a target to be buildable, with actually _building_ that target!  Fixing that wasn't hard, once I understood, but then the problem was figuring out a way to do it without hardcoding vcxproj in the build structure...  (which I did...)


Ongoing work here:  https://github.com/rec/rippled/tree/x-scons-2

**Comment by seelabs** [comment]
Closing this, it is no longer being actively worked on. It may be re-opened later.
